sync-core/src/test/java/com/cloudant/sync/datastore/AttachmentStreamFactoryTest.java

```diff
@@ -0,0 +1,370 @@
+
+        Assert.assertTrue("Reading unencrypted, encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        savedAttachment.getInputStream(),
+                        new FileInputStream(plainText)));
+
+    }
+
+    // =====================
+    // Encrypted, un-encoded
+    // =====================
+
+    @Test
+    /**
+     * Assert passing an AttachmentStreamFactory with known key writes an encrypted file to disk.
+     */
+    public void testPreparedAttachmentWritesEncryptedUnencodedStream() throws AttachmentException,
+            IOException, InvalidKeyException {
+
+        AttachmentStreamFactory asf = new AttachmentStreamFactory(
+                EncryptionTestConstants.keyProvider16Byte
+        );
+
+        File plainText = f("fixture/EncryptedAttachmentTest_plainText");
+
+        UnsavedFileAttachment usf = new UnsavedFileAttachment(plainText, "text/plain");
+        PreparedAttachment preparedAttachment = new PreparedAttachment(
+                usf, datastore_manager_dir, asf);
+
+        // We check content using EncryptedAttachmentInputStream. It seems a bit circular,
+        // but the aim of this test is to make sure AttachmentStreamFactory is providing
+        // encrypted output readable by EncryptedAttachmentInputStream rather than exactly
+        // what that output is (EncryptedAttachmentOutputStream's own tests do that).
+        Assert.assertTrue("Writing to encrypted, un-encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        new EncryptedAttachmentInputStream((
+                                new FileInputStream(preparedAttachment.tempFile)),
+                                EncryptionTestConstants.key16Byte),
+                        new FileInputStream(plainText)));
+
+    }
+
+    @Test
+    /**
+     * Assert passing an AttachmentStreamFactory with known key reads an encrypted file correctly.
+     */
+    public void testSavedAttachmentReadsEncryptedUnencodedStream() throws AttachmentException, IOException {
+
+        AttachmentStreamFactory asf = new AttachmentStreamFactory(
+                EncryptionTestConstants.keyProvider16Byte
+        );
+
+        File plainText = f("fixture/EncryptedAttachmentTest_plainText");
+        File cipherAttachmentBlob = f("fixture/EncryptedAttachmentTest_cipherText_aes128");
+
+        SavedAttachment savedAttachment = new SavedAttachment("test", 0, 0, null, "text/plain",
+                cipherAttachmentBlob, Attachment.Encoding.Plain, asf);
+
+        Assert.assertTrue("Reading encrypted, un-encoded stream didn't give correct output",
+                IOUtils.contentEquals(
+                        savedAttachment.getInputStream(),
+                        new FileInputStream(plainText)));
+
+    }
+
+    @Test
+    /**
+     * Assert passing an AttachmentStreamFactory with known key to both a PreparedAttachment
```

Comment:
This comment doesn't really describe this test very well. What it's actually doing is checking that when the attachment is read with a different key from the one used to write it, it isn't correctly decrypted.